docs/CODE_GUIDELINES.md

```diff
@@ -280,82 +415,200 @@ enum Dummy
+
+✅ Good:
+```cpp
+/*!
+ * \brief Set decoration state and size by providing full surface size
+ *        including decorations
+ *
+ * Calculates size of the main surface from size of all surfaces combined (including
+ * window decorations) by subtracting the decoration size
+ *
+ * Decorations will be disabled if state includes STATE_FULLSCREEN
+ *
+ * Call only from main thread
+ */
+void SetState(CSizeInt size, int scale, IShellSurface::StateBitset state);
+```
+
+❌ Bad:
+```cpp
+/**
+ * @brief Function for documentation purposes (javadoc style)
+ */
+void TestFunction();
+
+void ReallyComplicatedFunction(); // does something really complicated
+
+/*!
+ * \brief Frobnicate a parameter
+ * \param param parameter to frobnicate
+ * \result frobnication result
+ */
+int Frobnicate(int param);
 ```
 
 **[back to top](#table-of-contents)**
 
-## 12. Conventions
-### 12.1. Casts
-New code has to use C++ style casts and not older C style casts. When modifying existing code the developer can choose to update it to C++ style casts or leave as is. Remember that whenever a `dynamic_cast` is used on a pointer object the result can be a `nullptr` and needs to be checked accordingly.
+## 10. Other conventions
+### 10.1. Destructors in interfaces
+
+A class with any virtual functions should have a destructor that is either public and virtual or else protected and nonvirtual (cf. [ISO C++ guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-virtual)).
+
+### 10.2. Output parameters
+For functions that have multiple output values, prefer using a `struct` or `tuple` return value over output parameters that use pointers or references (cf. [ISO C++ guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out-multi)). In general, try to avoid output parameters completely (cf. [ISO C++ guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out), [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Output_Parameters)). At the function call site, it is completely invisible that actually a reference is being passed and the value might be modified. Return semantics make it clear what is happening.
+
+### 10.3. Casts
+New code has to use C++ style casts and not older C style casts. When modifying existing code the developer can choose to update it to C++ style casts or leave as is. Whenever a `dynamic_cast` is used to cast to a pointer type, the result can be `nullptr` and needs to be checked accordingly.
 
-### 12.2. NULL vs nullptr
+### 10.4. `NULL` vs `nullptr`
 Prefer the use of `nullptr` instead of `NULL`. `nullptr` is a typesafe version and as such can't be implicitly converted to `int` or anything else.
 
-### 12.3. auto
+### 10.5. `auto`
 
 Feel free to use `auto` wherever it improves readability. Good places are iterators or when dealing with containers.
```

Comment:
We should discourage `auto` when it isn't necessary for readability, as it is opaque compared to using the actual types.

Reply:
I prefer using auto whenever the type is obvious from context

Reply:
With @garbear on this one

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-auto
> Exception: Avoid auto for initializer lists and in cases where you know exactly which type you want and where an initializer might require conversion.

https://google.github.io/styleguide/cppguide.html#auto
> Use auto to avoid type names that are noisy, obvious, or unimportant - cases where the type doesn't aid in clarity for the reader. Continue to use manifest type declarations when it helps readability.

But maybe that's actually also what you mean with "obivous from context" @Paxxi ?

Reply:
Yeah, if a method is named `GetFileItems` I don't need to spell out that it's a `vector<CFileItem>`

Reply:
What's obvious to you may not be obvious to a new dev unfamiliar with the code.

Reply:
It isn't about knowing that size_t is 64bit, it's about knowing that vec.size() returns a size_t. Obviously most devs know this, but being explicit is explanatory to new devs.

Reply:
That is true. I don't believe it to be an issue in practice though. And when in doubt your ide will display the type on hover or similar method.
I'm not sure it makes sense to force a style optimized for simple text editor users.
Either don't mention auto or say something like, use your judgment and when in doubt write the type

Reply:
Also with garbear on readability (I read with my eyes, not pointing the mouse)